File: packages/elixir/lumis/test/annotations_test.exs
defmodule Lumis.AnnotationsTest do
  use ExUnit.Case, async: false

  import ExUnit.CaptureIO

  alias Lumis.Annotation

  defmodule TestFormatter do
    @behaviour Lumis.Formatter

    @impl true
    def render(source, events, _options) do
      Enum.map(events, fn
        {:start, %{scope: "punctuation.bracket.rainbow." <> _level}} ->
          send(self(), :saw_rainbow_bracket)
          []

        {:source, %{start: start, end: end_offset}} ->
          binary_part(source, start, end_offset - start)

        {:annotation_start, %Annotation{range: range, data: %{id: id}}} ->
          send(self(), {:resolved_range, id, range})
          ["<annotation:", Integer.to_string(id), ">"]

        :annotation_end ->
          "</annotation>"

        _syntax_event ->
          []
      end)
    end
  end

  test "annotations and their Elixir data reach a custom formatter" do
    source = "(price + tax)"

    annotations = [
      [offset: {1, byte_size(source) - 1}, data: %{id: 7}]
    ]

    assert {:ok, output} =
             Lumis.highlight(source,
               formatter: {TestFormatter, language: "elixir"},
               annotations: annotations,
               rainbow_brackets: true
             )

    assert output == "(<annotation:7>price + tax</annotation>)"
    assert_received {:resolved_range, 7, {1, 12}}
    assert_received :saw_rainbow_bracket
  end

  test "invalid UTF-8 byte boundaries are rejected against the source" do
    annotations = [
      [offset: {1, 2}, data: %{}]
    ]

    assert_raise Lumis.HighlightError, ~r/not a UTF-8 character boundary/, fn ->
      Lumis.highlight!("π",
        formatter: {TestFormatter, language: "elixir"},
        annotations: annotations
      )
    end
  end

  test "position ranges use zero-based UTF-8 byte columns" do
    source = "π\ncafé"

    annotations = [
      [position: {{1, 0}, {1, 5}}, data: %{id: 8}]
    ]

    assert {:ok, output} =
             Lumis.highlight(source,
               formatter: {TestFormatter, language: "elixir"},
               annotations: annotations
             )

    assert output == "π\n<annotation:8>café</annotation>"
    assert_received {:resolved_range, 8, {3, 8}}
  end

  test "position columns must be UTF-8 byte boundaries" do
    annotations = [
      [position: {{1, 0}, {1, 4}}, data: %{}]
    ]

    assert_raise Lumis.HighlightError, ~r/not a UTF-8 character boundary/, fn ->
      Lumis.highlight!("π\ncafé",
        formatter: {TestFormatter, language: "elixir"},
        annotations: annotations
      )
    end
  end

  defp highlight_with(annotations) do
    Lumis.highlight("a\n\nb", formatter: :html_inline, annotations: annotations)
  end

  test "rejects a range that runs backwards" do
    assert_raise NimbleOptions.ValidationError,
                 ~r/annotation 0 offset range start must not be after its end: 6\.\.4/,
                 fn -> highlight_with([[offset: {6, 4}]]) end

    assert_raise NimbleOptions.ValidationError,
                 ~r/annotation 0 position range start must not be after its end/,
                 fn ->
                   highlight_with([
                     [position: {{2, 0}, {1, 0}}]
                   ])
                 end
  end

  test "accepts an empty range as a point" do
    assert {:ok, _html} = highlight_with([[offset: {1, 1}]])

    assert {:ok, _html} =
             highlight_with([[position: {{0, 1}, {0, 1}}]])
  end

  test "requires exactly one of :offset and :position" do
    assert_raise NimbleOptions.ValidationError,
                 ~r/annotation 0 needs an :offset or a :position range/,
                 fn -> highlight_with([[data: %{}]]) end

    assert_raise NimbleOptions.ValidationError,
                 ~r/annotation 0 sets both :offset and :position, which is ambiguous/,
                 fn ->
                   highlight_with([
                     [offset: {0, 1}, position: {{0, 0}, {0, 1}}]
                   ])
                 end
  end

  test "rejects an annotation that is not a keyword list" do
    assert_raise NimbleOptions.ValidationError,
                 ~r/annotation 0 must be a keyword list with :offset or :position/,
                 fn -> highlight_with([%{offset: {0, 1}}]) end
  end

  test "rejects an unknown key rather than ignoring it" do
    assert_raise NimbleOptions.ValidationError,
                 ~r/annotation 0: unknown options \[:colour\]/,
                 fn ->
                   highlight_with([[offset: {0, 1}, colour: :red]])
                 end
  end

  test "rejects a malformed position" do
    assert_raise NimbleOptions.ValidationError,
                 ~r/annotation 0: invalid tuple in :position option/,
                 fn ->
                   highlight_with([[position: {{0}, {0, 1}}]])
                 end
  end

  test "data defaults to nil when omitted" do
    defmodule DataFormatter do
      @behaviour Lumis.Formatter

      @impl true
      def render(_source, events, _options) do
        for {:annotation_start, annotation} <- events, do: inspect(annotation.data)
      end
    end

    assert {:ok, "nil"} =
             Lumis.highlight("ab",
               formatter: {DataFormatter, language: "elixir"},
               annotations: [[offset: {0, 1}]]
             )
  end

  test "a point annotation on a blank line reaches the formatter" do
    source = "a\n\nb"

    annotations = [[offset: {2, 2}, data: %{id: 3}]]

    assert {:ok, output} =
             Lumis.highlight(source,
               formatter: {TestFormatter, language: "elixir"},
               annotations: annotations
             )

    assert output == "a\n<annotation:3></annotation>\nb"
    assert_received {:resolved_range, 3, {2, 2}}
  end

  test "the example renders every annotation shape" do
    example = Path.expand("../examples/annotations.exs", __DIR__)
    output = capture_io(fn -> Code.require_file(example) end)

    # A position range crossing a line boundary.
    assert output =~ ~s(<span class="line-changed">)
    # An offset range starting mid-token splits the `variable` scope.
    assert output =~
             ~s(<span class="l-variable">p</span><mark class="edit"><span class="l-variable">rice</span>)

    # Byte offsets over a multibyte literal.
    assert output =~ "&quot;☕ café&quot;"
    # A point renders as an empty element.
    assert output =~ ~s(<i data-note="why the gap?"></i>)
    # The overlapping annotation is closed and reopened, so it starts twice.
    assert length(String.split(output, ~s(<mark class="right">))) - 1 == 2
  end
end

File: packages/javascript/lumis/test/annotations.test.ts
import { beforeAll, describe, expect, it } from "vitest";

import javascript from "../langs/javascript.ts";
import { createHighlighter, type Annotation } from "../src/index.js";
import type { Formatter } from "../src/formatters.js";
import { renderExample } from "../examples/annotations.ts";
import { configureLocalWasmResolver } from "./wasm.js";

interface Change {
  id: number;
}

describe("annotations", () => {
  beforeAll(() => {
    configureLocalWasmResolver(["javascript"]);
  }, 120_000);

  it("flows typed annotations through the public highlight options", async () => {
    const source = "const total = (price + tax)";
    const changedText = "price + tax";
    const start = new TextEncoder().encode(source.slice(0, source.indexOf(changedText))).length;
    const annotations: Annotation<Change>[] = [
      {
        range: { type: "offset", start, end: start + new TextEncoder().encode(changedText).length },
        data: { id: 7 },
      },
    ];
    const formatter: Formatter<Change> = {
      language: javascript,
      render(rendered, events) {
        const bytes = new TextEncoder().encode(rendered);
        const decoder = new TextDecoder();
        const output: string[] = [];

        for (const event of events) {
          if (event.type === "start") output.push(`<syntax:${event.scope}>`);
          if (event.type === "end") output.push("</syntax>");
          if (event.type === "annotationStart") {
            output.push(`<annotation:${event.annotation.data.id}>`);
          }
          if (event.type === "annotationEnd") output.push("</annotation>");
          if (event.type === "source") {
            output.push(decoder.decode(bytes.subarray(event.start, event.end)));
          }
        }

        return output.join("");
      },
    };
    const highlighter = await createHighlighter({ languages: [javascript] });

    const output = highlighter.highlight(source, formatter, {
      annotations,
      rainbowBrackets: true,
    });

    expect(output).toContain("<syntax:");
    expect(output).toContain("<syntax:punctuation.bracket.rainbow.");
    expect(output).toContain("<annotation:7>");
    expect(output).toContain("</annotation>");
    expect(output.replaceAll(/<[^>]+>/g, "")).toBe(source);
  }, 30_000);

  it("rejects annotation offsets that are not UTF-8 boundaries", async () => {
    const source = "const café = 1";
    const annotations: Annotation[] = [
      {
        range: { type: "offset", start: 10, end: 11 },
        data: {},
      },
    ];
    const formatter: Formatter = {
      language: javascript,
      render(_source, events) {
        return JSON.stringify(events);
      },
    };
    const highlighter = await createHighlighter({ languages: [javascript] });

    expect(() => highlighter.highlight(source, formatter, { annotations })).toThrow(
      "is not a UTF-8 character boundary",
    );
  }, 30_000);

  it("resolves position ranges to UTF-8 byte ranges before formatting", async () => {
    const source = "const π = 3\nconst café = 4";
    let resolvedRange: { start: number; end: number } | undefined;
    const annotations: Annotation<Change>[] = [
      {
        range: {
          type: "position",
          start: { line: 1, column: 6 },
          end: { line: 1, column: 11 },
        },
        data: { id: 8 },
      },
    ];
    const formatter: Formatter<Change> = {
      language: javascript,
      render(rendered, events) {
        const sourceBytes = new TextEncoder().encode(rendered);
        const decoder = new TextDecoder();

        return events
          .map((event) => {
            if (event.type === "annotationStart") {
              resolvedRange = event.annotation.range;
            }
            if (event.type === "source") {
              return decoder.decode(sourceBytes.subarray(event.start, event.end));
            }
            return "";
          })
          .join("");
      },
    };
    const highlighter = await createHighlighter({ languages: [javascript] });

    const output = highlighter.highlight(source, formatter, { annotations });
    const start = new TextEncoder().encode(source.slice(0, source.indexOf("café"))).length;

    expect(output).toBe(source);
    expect(resolvedRange).toEqual({ start, end: start + new TextEncoder().encode("café").length });
  }, 30_000);

  it("rejects position columns that are not UTF-8 boundaries", async () => {
    const source = "π\ncafé";
    const annotations: Annotation[] = [
      {
        range: {
          type: "position",
          start: { line: 1, column: 0 },
          end: { line: 1, column: 4 },
        },
        data: {},
      },
    ];
    const formatter: Formatter = {
      language: javascript,
      render(_source, events) {
        return JSON.stringify(events);
      },
    };
    const highlighter = await createHighlighter({ languages: [javascript] });

    expect(() => highlighter.highlight(source, formatter, { annotations })).toThrow(
      "is not a UTF-8 character boundary",
    );
  }, 30_000);

  it("marks a blank line with an empty annotation", async () => {
    // The line-based review case: a blank line has nothing to cover, but it is
    // still somewhere a comment can land.
    const source = "const a = 1;\n\nconst b = 2;";
    const blankLineOffset = source.indexOf("\n") + 1;
    const annotations: Annotation<Change>[] = [
      {
        range: { type: "offset", start: blankLineOffset, end: blankLineOffset },
        data: { id: 1 },
      },
    ];
    const formatter: Formatter<Change> = {
      language: javascript,
      render(rendered, events) {
        const bytes = new TextEncoder().encode(rendered);
        const decoder = new TextDecoder();
        const parts: string[] = [];
        for (const event of events) {
          if (event.type === "annotationStart") parts.push(`<mark:${event.annotation.data.id}>`);
          else if (event.type === "annotationEnd") parts.push("</mark>");
          else if (event.type === "source") {
            parts.push(decoder.decode(bytes.subarray(event.start, event.end)));
          }
        }
        return parts.join("");
      },
    };
    const highlighter = await createHighlighter({ languages: [javascript] });

    expect(highlighter.highlight(source, formatter, { annotations })).toBe(
      "const a = 1;\n<mark:1></mark>\nconst b = 2;",
    );
  }, 30_000);

  it("does not let a point annotation leak into the rest of the document", async () => {
    const source = "const a = 1;";
    const annotations: Annotation<Change>[] = [
      { range: { type: "offset", start: 5, end: 5 }, data: { id: 2 } },
    ];
    const formatter: Formatter<Change> = {
      language: javascript,
      render(_rendered, events) {
        const starts = events.filter((event) => event.type === "annotationStart").length;
        const ends = events.filter((event) => event.type === "annotationEnd").length;
        return `${starts}:${ends}`;
      },
    };
    const highlighter = await createHighlighter({ languages: [javascript] });

    expect(highlighter.highlight(source, formatter, { annotations })).toBe("1:1");
  }, 30_000);

  it("rejects a range that runs backwards", async () => {
    const source = "const a = 1;";
    const annotations: Annotation<Change>[] = [
      { range: { type: "offset", start: 8, end: 5 }, data: { id: 3 } },
    ];
    const formatter: Formatter<Change> = {
      language: javascript,
      render: (_rendered, events) => String(events.length),
    };
    const highlighter = await createHighlighter({ languages: [javascript] });

    expect(() => highlighter.highlight(source, formatter, { annotations })).toThrow(
      "must not be after its end",
    );
  }, 30_000);

  it("renders every annotation shape in the example", async () => {
    const output = await renderExample();

    // A position range crossing a line boundary.
    expect(output).toContain('<span class="line-changed">');
    // An offset range starting mid-token splits the `variable` scope.
    expect(output).toContain(
      '<span class="l-variable">p</span><mark class="edit"><span class="l-variable">rice</span>',
    );
    // Byte offsets over a multibyte literal.
    expect(output).toContain("&quot;☕ café&quot;");
    // A point renders as an empty element.
    expect(output).toContain('<i data-note="why the gap?"></i>');
    // The overlapping annotation is closed and reopened, so it starts twice.
    expect(output.match(/<mark class="right">/g)).toHaveLength(2);
  }, 30_000);
});
